File: 333. Largest BST Subtree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.min = -sys.maxsize-1
        self.max = sys.maxsize
    def largestBSTSubtree(self, root: 'TreeNode') -> 'int':
        if not root:
            return 0
        
        _, _, count = self.myLargestBSTSubtree(root)
        return count
    
    def myLargestBSTSubtree(self, root):
        if root.right and root.left:
            llow, lhigh, lcount = self.myLargestBSTSubtree(root.left)
            rlow, rhigh, rcount = self.myLargestBSTSubtree(root.right)
            if llow <= lhigh < root.val < rlow <= rhigh:
                return llow, rhigh, lcount + rcount + 1
            else:
                if lcount > rcount:
                    return self.max, self.min, lcount
                else:
                    return self.max, self.min, rcount
        elif not root.right and root.left:
            llow, lhigh, lcount = self.myLargestBSTSubtree(root.left)
            if llow <= lhigh < root.val:
                return llow, root.val, lcount + 1
            else:
                return self.max, self.min, lcount
        elif root.right and not root.left:
            rlow, rhigh, rcount = self.myLargestBSTSubtree(root.right)
            if root.val < rlow <= rhigh:
                return root.val, rhigh, rcount + 1
            else:
                return self.max, self.min, rcount
        else:
            return root.val, root.val, 1
                    

File: 333. Largest BST Subtree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int largestBSTSubtree(TreeNode root) {
        SubTree subTree = helper(root);
        return subTree.count;
    }
    
    public SubTree helper(TreeNode root) {
        SubTree left, right, subTree;
        if (root == null) {
            return new SubTree(0, -1, -1, true);
        }
        

        left = helper(root.left);
        right = helper(root.right);
        int min = (left.min == -1) ? root.val : left.min;
        int max = (right.max == -1) ? root.val : right.max;
        if ((root.left == null || root.val > left.max) && 
            (root.right == null || root.val < right.min) && 
            left.isBST && right.isBST) {
            return new SubTree(left.count + right.count + 1, min, max, true);
        } else {
            return new SubTree(Math.max(left.count, right.count), min, max, false);
        }
    }
    
    public class SubTree {
        public int count;
        public boolean isBST;
        public int min;
        public int max;
        
        public SubTree(int count, int min, int max, boolean isBST) {
            this.count = count;
            this.min = min;
            this.max = max;
            this.isBST = isBST;
        }
    }
}
